article/results.tex

```diff
@@ -56,59 +56,75 @@ \subsection{Best Practice Guidelines}
+This can be accomplished in POSIX shell script via using \texttt{cd \textquotedbl\$(dirname \textquotedbl\$0\textquotedbl)\textquotedbl} at the beginning of the script.
 
-\subsubsection{Hierarchical granularity greatly benefits efficiency}
+\subsubsection{Workflow granularity greatly benefits efficiency}
 The high time cost of executing the full analysis workflow makes debugging errors very time-consuming.
 Ideally it should not be necessary to re-execute the entire workflow once an error is potentially resolved.
-For this it is beneficial to segment the workflow into as many hierarchical steps as feasible.
-This is not always possible, and is particularly not feasible for higher-level statistical analysis, where hierarchical contingency can easily be lost.
-However, particularly for such steps as preprocessing, and first-level general linear modelling, this is commonly feasible, and should be done.
-An easy implementation of this is having the workflow coordination system check for the presence of each hierarchical result, and, if present, proceed to the next hierarchical step.
+For this it is beneficial to formulate the workflow as many separate steps, where steps could be troubleshooted independently.
+It is worth to separate out at least such large steps as preprocessing, individual levels of analysis (e.g. per-subject vs whole-population), and article generation.
+YODA principles~\cite{yoda} again demonstrate one way how to organize digital artifacts across such stages for maximal re-usability.
+An good feature is to use a workflow platform which automatically checks for the presence of results from prior stages, and, if present, proceed to the next stage without triggering re-computation of prior stages.
+In this article we rely on simple GNU Make rules to provide very basic workflow across computational stages, but there exist hundreds of workflow engines and we refer reader to a recent call for FAIR workflows to discover more~\cite{fair-workflows}.
 
 \subsubsection{Container image size should be kept small}
 Due to a lack of persistency, addressing issues in container image contents requires rebuilding, which can be a time-consuming process.
 The smaller the container image, the easier it is.
 In particular, when using containers, it is thus advisable to \textit{not} provide data via a package manager or via manual download inside the build script.
-A suitable alternative is to assure provision of these resources outside of the container image, e.g. by bind-mounting Git (and DataLad) directories present on the host machine.
+A suitable alternative is to assure provision of these resources outside of the container image, e.g. by bind-mounting Git (and DataLad) directories present on the host machine or even combining them all via Git submodules mechanism as promoted by YODA principles~\cite{yoda}.
```

Comment:
I think “combining” here is a bit confusing. Also, is it really related to image size? I'm not sure whether YODA is related to this point.

Reply:
it relates to composition... please suggest alternative to "combining" to make it less confusing

Reply:
I can't because I'm confused about what you are describing. It doesn't matter with respect to image size whether they're modules or submodules. Just that they're in a directory which can be bind-mounted. Or am I missing something?

Reply:
note -- containers nicely bind things together...
But we are saying: "do not include large data into containers, just bind from outside".  But binding from outside "breaks" that binding.  Here I am promoting combining with `git submodules` so it binds it all back together .
I agree that such line of thinking is not obvious from what is written though. See how to improve?

Reply:
```suggestion
A suitable alternative is to assure provision of these resources outside of the container image (e.g. via Git and Datalad) and bind-mounting data present on the host machine.
```

This is an improvement I think, but it doesn't actually include your point. I think you're trying to make a separate point. We already say “assure provision”. Whether the repository uses `wget ... && tar xvf ...` or `datalad install && datalad get ...` or whether these are submodules or just copies where git/datalad are only used as an explicit download command I think is irrelevant to the point of keeping images small.

Reply:
```suggestion
A suitable alternative is to assure provision of these resources outside of the container image and bind-mounting data thus downloaded to the host machine.

\subsubsection{Resources  should be bundled into a DataLad superdataset}
As external resources might change or disappear, it is beneficial to use data version control system, such as git-annex and DataLad.
The git submodule mechanism allows to bundle multiple repositories together with clear versioning information, thus following the modularity principle promoted by YODA~\cite{yoda}.
Moreover, git-annex allows for multiple data sources and data integrity verification.
```